server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistoryService.java

```diff
@@ -60,14 +79,23 @@ public MasterHistory getLocalMasterHistory() {
      * updated even if the ClusterState is updated on this node or the remote node. The history is retrieved asynchronously, and only if
      * requestRemoteMasterHistory has been called for this node. If anything has gone wrong fetching it, the exception returned by the
      * remote machine will be thrown here. If the remote history has not been fetched or if something went wrong and there was no exception,
-     * the returned value will be null.
+     * the returned value will be null. If the remote history is old enough to be considered stale (that is, older than
+     * MAX_USABLE_REMOTE_HISTORY_AGE_SETTING), then the returned value will be null.
      * @return The MasterHistory from a remote node's point of view. This MasterHistory object will not be updated with future changes
      * @throws Exception the exception (if any) returned by the remote machine when fetching the history
      */
     @Nullable
     public List<DiscoveryNode> getRemoteMasterHistory() throws Exception {
         // Grabbing a reference to the object in case it is replaced in another thread during this method:
         RemoteHistoryOrException remoteHistoryOrExceptionCopy = remoteHistoryOrException;
+        /*
+         * If the remote history we have is too old, we just return null with the assumption that it is stale and the new one has not
+         * come in yet.
+         */
+        long acceptableRemoteHistoryTime = currentTimeMillisSupplier.getAsLong() - acceptableRemoteHistoryAge.getMillis();
+        if (remoteHistoryOrExceptionCopy.creationTimestamp < acceptableRemoteHistoryTime) {
+            return null;
```

Comment:
Would we want to also at some point null-ify `remoteHistoryOrException` ?  (we'd likely need an AtomicReference or such to do a compare and swap) 

What do you think?

Reply:
My thinking was that there was no harm in holding onto a single one, and it avoids having to synchronize access to it. Synchronizing it would not be super expensive (we'd need to read it, do the time calculation, and set it to null in an atomic action), but it seemed like an unnecessary complication. Is it the memory that you're concerned about? We cap these things to have at most 50 entries, so I'd expect at most a few KB here.

Reply:
I was mostly worried about the `remoteHistoryOrException` leaking outside the service by other means than `getRemoteMasterHistory` or the service evolving and using the `remoteHistoryOrException` in a private way (assuming it's `null` when it's stale, which would not be a wild assumption given what the getter returns) - @DaveCTurner would you have a strong opinion here?

Reply:
I am inclined to leave it in place too. You never know, it might even end up being useful in a heap dump. Maybe add a comment to the field indicating that it might be stale.

For future reference it wouldn't be a big deal to clear it after the timeout either:

```
        transportService.getThreadPool().schedule(() -> {/* clear field*/}, acceptableRemoteHistoryAge, Names.SAME);
```

(also there's no need for a `ThreadPool` constructor argument if you have a `TransportService`)

Reply:
++ let's leave it as is then